internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -36,12 +42,65 @@ func CreateInMemoryWriteBuffer() *InMemoryWriteBuffer {
 }
 
 func (b *InMemoryWriteBuffer) InitializeBuffer(sizeInMB int) {
-	ChunkSize = sizeInMB * MiB
+	ChunkSize = int64(sizeInMB * MiB)
+	BufferSize = 2 * ChunkSize
+	b.bufferEndOffset = ChunkSize
 	if b.buffer == nil {
-		b.buffer = bytes.NewBuffer(make([]byte, 0, 2*sizeInMB*MiB))
+		b.buffer = make([]byte, 0, BufferSize)
 	}
 }
 
-func (b *InMemoryWriteBuffer) WriteAt(data []byte, offset int64) error {
+func (b *InMemoryWriteBuffer) WriteAt(data []byte, dataStartOffset int64) error {
+	dataSize := int64(len(data))
+	dataEndOffset := dataStartOffset + dataSize
+
+	// If dataStartOffset is not in the range of offsets stored in buffer block,
+	// trigger temp file flow for writes.
+	if dataStartOffset < b.bufferStartOffset || b.bufferEndOffset < dataStartOffset {
+		// TODO: finalise write and trigger temp file flow.
+		return fmt.Errorf(NonSequentialWriteError)
+	}
+
+	// If data can be completely written without wrapping to previously written
+	// buffer block (beyond the current block of buffer), write it.
+	if dataEndOffset <= b.bufferEndOffset {
+		return b.copyDataToBuffer(dataStartOffset, dataEndOffset, data)
+	} else {
+		// Below logic is written with the assumption that the data received in a
+		// single write call from the kernel will never exceed beyond buffer block size.
+
+		// Write the chunk of data that can be written to the current buffer block.
+		l := b.bufferEndOffset - dataStartOffset
+		if err := b.copyDataToBuffer(dataStartOffset, b.bufferEndOffset, data[:l]); err != nil {
+			return err
+		}
+
+		//TODO: get status of last write to GCS, and trigger async write for current block.
+
+		// Update buffer offsets for next buffer block.
+		b.bufferStartOffset = b.bufferEndOffset
+		b.bufferEndOffset += ChunkSize
+		// Write the remaining data by wrapping over to the next buffer block.
+		return b.copyDataToBuffer(b.bufferStartOffset, dataEndOffset, data[l:])
+	}
+}
+
+func (b *InMemoryWriteBuffer) copyDataToBuffer(dataStartOffset, dataEndOffset int64, data []byte) error {
+	dataSize := int64(len(data))
+	si := dataStartOffset % BufferSize
```

Comment:
Shouldn't this be 

```suggestion
	si := dataStartOffset % ChunkSize
```

? Or we need to track which chunk (first half or second half) of the buffer to write to..

Reply:
This is obsolete now as we use 2 buffers